Add from_timestamp_micros function

umanwizard opened the pull request:
Fixes #904

djc commented on src/naive/datetime/tests.rs:
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
+        (2034061609000000, "2034-06-16 09:06:49.000000000"),
+    ];
+
+    for (timestamp_micros, formatted) in valid_map.iter().cloned() {
```
Nit: `cloned()` -> `copied()`.

djc commented on src/naive/datetime/tests.rs:
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
+
+    let invalid = [i64::MAX, i64::MIN];
+
+    for timestamp_micros in invalid.iter().cloned() {
```
Same here.

djc commented on src/naive/datetime/tests.rs:
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
+    // Test that the result of `from_timestamp_micros` compares equal to
+    // that of `from_timestamp_opt`.
+    let secs_test = [0, 1, 2, 1000, 1234, 12345678, -1, -2, -1000, -12345678];
+    for secs in secs_test.iter().cloned() {
```
And here.

djc commented on src/naive/datetime/mod.rs:
```diff
@@ -129,6 +155,36 @@ impl NaiveDateTime {
         datetime.expect("invalid or out-of-range datetime")
     }
 
+    /// Creates a new [NaiveDateTime] from milliseconds or microseconds since the UNIX epoch.
```
Style nit: I would prefer this (private) API to be ordered below the public users of this method.

djc reviewed (approved):
LGTM, thanks!

umanwizard commented:
I have changed `cloned` to `copies` in the test (and in the similar test for `from_timestamp_millis`), and moved the private method to the very bottom of the impl block.

esheppa commented on src/naive/datetime/tests.rs:
```diff
@@ -17,20 +17,57 @@ fn test_datetime_from_timestamp_millis() {
+
+    // Test that the result of `from_timestamp_micros` compares equal to
+    // that of `from_timestamp_opt`.
+    let secs_test = [0, 1, 2, 1000, 1234, 12345678, -1, -2, -1000, -12345678];
```
Thanks for adding this here as well!

esheppa reviewed (approved):
Looks great!
